File: submission/ngawande/tests/helpers/db_helpers.py
"""Database query helpers — direct DB assertions separated from test logic."""

from __future__ import annotations

import json
import sqlite3


def get_wallet_balance(db: sqlite3.Connection, wallet_id: str) -> int:
    """Return the current balance of a wallet."""
    row = db.execute(
        "SELECT balance FROM wallets WHERE id = ?", (wallet_id,)
    ).fetchone()
    assert row is not None, f"Wallet {wallet_id} not found"
    return row["balance"]


def get_transfer_count(db: sqlite3.Connection) -> int:
    """Return total number of transfer rows."""
    return db.execute("SELECT COUNT(*) FROM transfers").fetchone()[0]


def get_transfer_by_id(db: sqlite3.Connection, transfer_id: str) -> dict | None:
    """Return a transfer row as a dict, or None."""
    row = db.execute(
        "SELECT * FROM transfers WHERE id = ?", (transfer_id,)
    ).fetchone()
    return dict(row) if row else None


def get_transfers_by_idempotency_key(
    db: sqlite3.Connection, key: str
) -> list[dict]:
    """Return all transfer rows matching an idempotency key."""
    rows = db.execute(
        "SELECT * FROM transfers WHERE idempotency_key = ?", (key,)
    ).fetchall()
    return [dict(r) for r in rows]


def get_audit_event_count(db: sqlite3.Connection) -> int:
    """Return total number of audit_events rows."""
    return db.execute("SELECT COUNT(*) FROM audit_events").fetchone()[0]


def get_audit_events_for_transfer(
    db: sqlite3.Connection, transfer_id: str
) -> list[dict]:
    """Return all audit_events rows for a given transfer_id."""
    rows = db.execute(
        "SELECT * FROM audit_events WHERE transfer_id = ?", (transfer_id,)
    ).fetchall()
    return [dict(r) for r in rows]


def get_outbox_event_count(db: sqlite3.Connection) -> int:
    """Return total number of outbox_events rows."""
    return db.execute("SELECT COUNT(*) FROM outbox_events").fetchone()[0]


def get_outbox_events_for_transfer(
    db: sqlite3.Connection, transfer_id: str
) -> list[dict]:
    """Return all outbox_events rows for a given transfer_id."""
    rows = db.execute(
        "SELECT * FROM outbox_events WHERE transfer_id = ?", (transfer_id,)
    ).fetchall()
    return [dict(r) for r in rows]


def get_outbox_event_payload(db: sqlite3.Connection, transfer_id: str) -> dict | None:
    """Return parsed JSON payload from the outbox event for a transfer."""
    rows = get_outbox_events_for_transfer(db, transfer_id)
    if not rows:
        return None
    return json.loads(rows[0]["payload"])


def get_total_wallet_balance(db: sqlite3.Connection) -> int:
    """Return sum of all wallet balances (conservation check)."""
    row = db.execute("SELECT SUM(balance) FROM wallets").fetchone()
    return row[0] or 0

File: submission/solution-mohith-krishna/tests/helpers/dbHelpers.ts
import type Database from 'better-sqlite3';
import type { Wallet, Transfer, AuditEvent, OutboxEvent } from '../../src/service/types';

export class DbHelpers {
  constructor(private db: Database.Database) {}

  getWallet(walletId: string): Wallet | undefined {
    return this.db.prepare('SELECT id, balance, currency FROM wallets WHERE id = ?')
      .get(walletId) as Wallet | undefined;
  }

  getWalletBalance(walletId: string): number {
    const row = this.getWallet(walletId);
    if (!row) throw new Error(`Wallet ${walletId} not found`);
    return row.balance;
  }

  getTransfer(transferId: string): Transfer | undefined {
    return this.db.prepare(
      `SELECT id, source_wallet_id, destination_wallet_id, amount, currency,
              reference, status, idempotency_key, payload_hash, created_at
       FROM transfers WHERE id = ?`
    ).get(transferId) as Transfer | undefined;
  }

  getTransferCount(): number {
    const row = this.db.prepare('SELECT COUNT(*) as count FROM transfers').get() as { count: number };
    return row.count;
  }

  getTransfersByIdempotencyKey(key: string): Transfer[] {
    return this.db.prepare(
      `SELECT id, source_wallet_id, destination_wallet_id, amount, currency,
              reference, status, idempotency_key, payload_hash, created_at
       FROM transfers WHERE idempotency_key = ?`
    ).all(key) as Transfer[];
  }

  getAuditEvents(transferId: string): AuditEvent[] {
    return this.db.prepare(
      'SELECT id, transfer_id, event_type, payload, created_at FROM audit_events WHERE transfer_id = ?'
    ).all(transferId) as AuditEvent[];
  }

  getAuditEventCount(): number {
    const row = this.db.prepare('SELECT COUNT(*) as count FROM audit_events').get() as { count: number };
    return row.count;
  }

  getOutboxEvents(transferId: string): OutboxEvent[] {
    return this.db.prepare(
      'SELECT id, transfer_id, event_type, payload, status, created_at FROM outbox_events WHERE transfer_id = ?'
    ).all(transferId) as OutboxEvent[];
  }

  getOutboxEventCount(): number {
    const row = this.db.prepare('SELECT COUNT(*) as count FROM outbox_events').get() as { count: number };
    return row.count;
  }

  getAllTransfers(): Transfer[] {
    return this.db.prepare(
      `SELECT id, source_wallet_id, destination_wallet_id, amount, currency,
              reference, status, idempotency_key, payload_hash, created_at
       FROM transfers`
    ).all() as Transfer[];
  }

  getTotalBalanceForCurrency(currency: string): number {
    const row = this.db.prepare(
      'SELECT SUM(balance) as total FROM wallets WHERE currency = ?'
    ).get(currency) as { total: number };
    return row.total;
  }
}
